dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java

```diff
@@ -0,0 +1,455 @@
+
+        for (String configValuePart : nameList.split(",")) {
+            Optional<ModuleName> processedName = parseModuleName(configValuePart, configKey);
+            processedName.ifPresent(result::add);
+        }
+
+        return result;
+    }
+
+    /**
+     * Parses a String of app and/or module names into a list of {@code ModuleName}
+     * <p>
+     * Names must be in one of these formats:
+     * <ul>
+     * <li>appName</li>
+     * <li>appName/moduleName</li>
+     * </ul>
+     *
+     * @param nameList thane
+     * @param configKey the name of the config property holding the list (used for reporting errors)
+     * @return the parsed name or an empty Optional if it did not fit the format.
+     */
+    private static Optional<ModuleName> parseModuleName(String name, String configKey) {
+        Matcher m = CONFIG_VALUE_NAME_REFERENCE.matcher(name);
+
+        if (!m.matches()) {
+            Tr.warning(tc, MessageConstants.OPENAPI_MERGE_INVALID_NAME_CWWKO1666W, configKey, name);
+            return Optional.empty();
+        }
+
+        String appName = m.group(1).trim();
+        String moduleName = m.group(3);
+        if (moduleName != null) {
+            moduleName = moduleName.trim();
+        }
+        return Optional.of(new ModuleName(appName, moduleName));
+    }
+
+    @Override
+    public int getConfigListenerPriority() {
+        return 1;
+    }
+
+    private ConfigValues getConfigValues() {
+        synchronized (this) {
+            if (configValues == null) {
+                configValues = new ConfigValues();
+            }
+            return configValues;
+        }
+    }
+
+    private class ConfigValues {
+        protected boolean isAll = false;
+        protected boolean isFirst = false;
+        protected boolean serverxmlmode = false;
+        protected List<ModuleName> included;
+        protected List<ModuleName> excluded;
+
+        /**
+         * Configures a {@code ModuleSelectionConfig}
+         * <p>
+         * This will read and parse the {@value Constants#MERGE_INCLUDE_CONFIG} and {@value Constants#MERGE_EXCLUDE_CONFIG} config properties from mpConfig.
+         * <p>
+         * TODO doc what it reads from server.xml
+         * <p>
+         * If the config is invalid, this method will output warning messages but still return a usable result object.
+         *
+         * @return the module selection config
+         */
+        protected ConfigValues() {
+
+            Config configFromMPConfig = ConfigProvider.getConfig(ApplicationRegistryImpl.class.getClassLoader());
+
+            if (!ProductInfo.getBetaEdition()) {
+                serverxmlmode = false;
+            } else {
+                serverxmlmode = configFromServerXML.getIncludedAppsAndModules().isPresent() ||
+                                configFromServerXML.getIncludedAppsAndModules().isPresent() ||
+                                configFromServerXML.getConfigMode().isPresent();
```

Comment:
```suggestion
                serverxmlmode = configFromServerXML.getIncludedAppsAndModules().isPresent() ||
                                configFromServerXML.getExcludedAppsAndModules().isPresent() ||
                                configFromServerXML.getConfigMode().isPresent();
```

Though this code may go away anyway if you replace these three methods with one method.